Library/Homebrew/test/cask/artifact/shared_examples/uninstall_zap.rb

```diff
@@ -89,43 +91,65 @@
+        .and_return(instance_double("SystemCommand::Result", stdout: "loaded"))
 
-      FakeSystemCommand.expects_command(
-        sudo(%W[/sbin/kextunload -b #{kext_id}]),
-      )
+      expect(subject).to receive(:system_command!)
+        .with("/sbin/kextunload", args: ["-b", kext_id], sudo: true)
+        .and_return(instance_double("SystemCommand::Result"))
 
-      FakeSystemCommand.expects_command(
-        sudo(%W[/usr/sbin/kextfind -b #{kext_id}]), "/Library/Extensions/FancyPackage.kext\n"
-      )
+      expect(subject).to receive(:system_command!)
+        .with("/usr/sbin/kextfind", args: ["-b", kext_id], sudo: true)
+        .and_return(instance_double("SystemCommand::Result", stdout: "/Library/Extensions/FancyPackage.kext\n"))
 
-      FakeSystemCommand.expects_command(
-        sudo(["/bin/rm", "-rf", "/Library/Extensions/FancyPackage.kext"]),
-      )
+      expect(subject).to receive(:system_command!)
+        .with("/bin/rm", args: ["-rf", "/Library/Extensions/FancyPackage.kext"], sudo: true)
 
-      subject
+      subject.public_send(:"#{artifact_dsl_key}_phase", command: fake_system_command)
     end
   end
 
   context "using :quit" do
     let(:cask) { Cask::CaskLoader.load(cask_path("with-#{artifact_dsl_key}-quit")) }
     let(:bundle_id) { "my.fancy.package.app" }
-    let(:quit_application_script) do
-      %Q(tell application id "#{bundle_id}" to quit)
+
+    it "is skipped when the user is not a GUI user" do
+      allow(User.current).to receive(:gui?).and_return false
+      allow(subject).to receive(:running_processes).with(bundle_id).and_return([[0, "", bundle_id]])
+
+      expect {
+        subject.public_send(:"#{artifact_dsl_key}_phase", command: fake_system_command)
+      }.to output(/Not logged into a GUI; skipping quitting application ID 'my.fancy.package.app'\./).to_stdout
     end
 
-    it "is supported" do
-      FakeSystemCommand.stubs_command(
-        %w[/bin/launchctl list], "999\t0\t#{bundle_id}\n"
-      )
+    it "quits a running application" do
+      allow(User.current).to receive(:gui?).and_return true
 
-      FakeSystemCommand.stubs_command(
-        %w[/bin/launchctl list],
-      )
+      expect(subject).to receive(:running_processes).with(bundle_id).ordered.and_return([[0, "", bundle_id]])
+      expect(subject).to receive(:quit).with(bundle_id)
+                                       .and_return(instance_double("SystemCommand::Result", success?: true))
+      expect(subject).to receive(:running_processes).with(bundle_id).ordered.and_return([])
 
-      subject
+      expect {
+        subject.public_send(:"#{artifact_dsl_key}_phase", command: fake_system_command)
+      }.to output(/Application 'my.fancy.package.app' quit successfully\./).to_stdout
+    end
+
+    it "tries to quit the application for 10 seconds" do
+      allow(User.current).to receive(:gui?).and_return true
+
+      allow(subject).to receive(:running_processes).with(bundle_id).and_return([[0, "", bundle_id]])
```

Comment:
Your test is a much better approach than the end-to-end test code I had in mind!

For reference, the following script (which I’m going to abandon now) blocks the AppleScript command `tell application id "homebrew.cask.test.AppThatQuitsSlowly" to quit` for 15 seconds:

```
#!/usr/bin/python

import time
from AppKit import NSApplication, NSObject
from Foundation import NSBundle
from PyObjCTools import AppHelper

class AppThatQuitsSlowly:
    def start(self):
        info = NSBundle.mainBundle().infoDictionary()
        info['CFBundleIdentifier'] = 'homebrew.cask.test.AppThatQuitsSlowly'
        NSApplication.sharedApplication().setDelegate_(self._create_app_delegate().alloc().init())
        AppHelper.runEventLoop()

    def _create_app_delegate(self):
        class AppDelegate(NSObject):
            def applicationShouldTerminate_(self, notification):
                time.sleep(15)
                return True

        return AppDelegate

AppThatQuitsSlowly().start()
```

Just burying the script here in case it comes in handy for some future ad-hoc test 😄